### ISSUE TITLE
Reset-on-Exit/Reset-on-Crit

### ISSUE THREAD

**Issue by eelstork** [active-logic/activelogic-cs#43]
[Reset-on-Resume](https://github.com/active-logic/activelogic-cs/blob/master/Doc/Reference/RoR.md) (RoR) is a relatively new feature in Active Logic. Experiments show that Reset-on-Resume is not dependable enough, since it may cause issues when repeating tasks seamlessly (see #37)
Reset-on-Exit is safer, and may be implemented as follows:

```cs
roe DoSomething() => `cx() && Wait(0.5) && ... ;  
```

Or

```cs
roe DoSomething(){
    // some preliminary work
    return cx() && Once()?[ task ] ... ;   
}
```

Above syntax is primarily attractive because it does not require delimiters; other syntaxes are possible; this may come with a few desirable features:

- RoE decorators fail if no RoE context is provided (runtime check)
- Stateless control does not require a context, avoiding both syntax and performance overheads.
- `cx()` and `roe` are interdependent, so one cannot accidentally be used without another (compile check)

**Comment by eelstork** [comment]
Draft fix for this. Syntax is conservative:

```cs
status Task => roe[ ... ];
```

Looks thread-safe.

One possible alternative *only* requires a pseudo-type. Will put this on a branch later, if only for the hell of it.

**Comment by eelstork** [comment]
Looks like both RoR and RoE may be needed to ensure decorators reset when they are expected to.

**Comment by eelstork** [comment]
Remaining tasks:
- Assess behavior when resettable decorators are not included within a RoE block. For now may want this to remain optional, however if the result is unpredictable behavior, may need to be more strict.
- Do ordered composites need this?

**Comment by eelstork** [comment]
Dependable solution here is a reset-on-criterion (RoC) (and perhaps later, reset-on-condition) block. Reset-on-exit (RoE) is documented in BT literature, will work in many cases, but I am seeing edge cases where results are incorrect; keeping for now pending further testing/evaluation.

- reset blocks (RoC, RoE) implemented in 09181f1c229e3d1521e6a7a34f79becfe74bce47
- RoR covers forks missed by reset blocks (already implemented).
